Memory leak in function `is_scalar_with_conversion` (a static analysis report with a POC)

### Reproducing code example:

```py
import operator
import sys
import numpy as np

class Bar(str):
    index = 500
    def __index__(self):
        return self.index

a = np.array([1, 2])
b = Bar('1')

bef_ref = sys.getrefcount(b.index)  # b.index->ob_refcnt == 4

for i in range(1000):
    try:
        operator.pow(a, b)
    except:
        pass

assert sys.getrefcount(b.index) == bef_ref  # b.index->ob_refcnt == 1004
```

As the reference count linearly increases with the number of invocations, which indicates a memory leak.

### Error message:

The path provided by the static analyzer is as follows.

1. In condition branch starts from here.
https://github.com/numpy/numpy/blob/04ab04d93d4d7a4d241fe0ceb725436a8b6c8c2e/numpy/core/src/multiarray/number.c#L422

2. A new reference returned from `PyNumber_Index` and pointed to be `value`.
https://github.com/numpy/numpy/blob/04ab04d93d4d7a4d241fe0ceb725436a8b6c8c2e/numpy/core/src/multiarray/number.c#L423

3. Assume `value` is not `NULL`.
https://github.com/numpy/numpy/blob/04ab04d93d4d7a4d241fe0ceb725436a8b6c8c2e/numpy/core/src/multiarray/number.c#L425

3. Call `PyLong_AsSsize_t` with `value`, which does not steal the reference.
https://github.com/numpy/numpy/blob/04ab04d93d4d7a4d241fe0ceb725436a8b6c8c2e/numpy/core/src/multiarray/number.c#L431

4. Function returns on the normal branch. (the error branch is also the same case)
https://github.com/numpy/numpy/blob/04ab04d93d4d7a4d241fe0ceb725436a8b6c8c2e/numpy/core/src/multiarray/number.c#L437

5. The new reference pointed to by `value` leaks without decreasing its refcnt.

### NumPy/Python version information:

Static analysis carried out on commit 04ab04d.

Oh wait, there's a second one, ending in
```
RuntimeError: Broken toolchain: cannot link a simple C program
```

<details>

```
Running from numpy source directory.
numpy/random\_bounded_integers.pxd.in has not changed
numpy/random\bit_generator.pyx has not changed
numpy/random\mtrand.pyx has not changed
numpy/random\_bounded_integers.pyx has not changed
numpy/random\_bounded_integers.pyx.in has not changed
numpy/random\_common.pyx has not changed
numpy/random\_generator.pyx has not changed
numpy/random\_mt19937.pyx has not changed
numpy/random\_pcg64.pyx has not changed
numpy/random\_philox.pyx has not changed
numpy/random\_sfc64.pyx has not changed
Cythonizing sources
blas_opt_info:
blas_mkl_info:
No module named 'numpy.distutils._msvccompiler' in numpy.distutils; trying from distutils
customize MSVCCompiler
  libraries mkl_rt not found in ['C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\lib', 'C:\\', 'C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\libs', 'C:\\Users\\nikol\\anaconda3\\Library\\lib']
  NOT AVAILABLE

blis_info:
  libraries blis not found in ['C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\lib', 'C:\\', 'C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\libs', 'C:\\Users\\nikol\\anaconda3\\Library\\lib']
  NOT AVAILABLE

openblas_info:
  libraries openblas not found in ['C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\lib', 'C:\\', 'C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\libs', 'C:\\Users\\nikol\\anaconda3\\Library\\lib']
get_default_fcompiler: matching types: '['gnu', 'intelv', 'absoft', 'compaqv', 'intelev', 'gnu95', 'g95', 'intelvem', 'intelem', 'flang']'
customize GnuFCompiler
Could not locate executable g77
Could not locate executable f77
customize IntelVisualFCompiler
Could not locate executable ifort
Could not locate executable ifl
customize AbsoftFCompiler
Could not locate executable f90
customize CompaqVisualFCompiler
Found executable C:\cyg\cygwin64\bin\DF.exe
customize IntelItaniumVisualFCompiler
Could not locate executable efl
customize Gnu95FCompiler
Could not locate executable gfortran
Could not locate executable f95
customize G95FCompiler
Could not locate executable g95
customize IntelEM64VisualFCompiler
customize IntelEM64TFCompiler
Could not locate executable efort
Could not locate executable efc
customize PGroupFlangCompiler
Found executable C:\Users\nikol\anaconda3\envs\numpy-dev\Library\bin\flang.exe
  NOT AVAILABLE

accelerate_info:
  NOT AVAILABLE

atlas_3_10_blas_threads_info:
Setting PTATLAS=ATLAS
  libraries tatlas not found in ['C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\lib', 'C:\\', 'C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\libs', 'C:\\Users\\nikol\\anaconda3\\Library\\lib']
  NOT AVAILABLE

atlas_3_10_blas_info:
  libraries satlas not found in ['C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\lib', 'C:\\', 'C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\libs', 'C:\\Users\\nikol\\anaconda3\\Library\\lib']
  NOT AVAILABLE

atlas_blas_threads_info:
Setting PTATLAS=ATLAS
  libraries ptf77blas,ptcblas,atlas not found in ['C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\lib', 'C:\\', 'C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\libs', 'C:\\Users\\nikol\\anaconda3\\Library\\lib']
  NOT AVAILABLE

atlas_blas_info:
  libraries f77blas,cblas,atlas not found in ['C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\lib', 'C:\\', 'C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\libs', 'C:\\Users\\nikol\\anaconda3\\Library\\lib']
  NOT AVAILABLE

C:\Users\nikol\OneDrive\Documents\OpenProjects\Python\numpy\numpy\distutils\system_info.py:2026: UserWarning: 
    Optimized (vendor) Blas libraries are not found.
    Falls back to netlib Blas library which has worse performance.
    A better performance should be easily gained by switching
    Blas library.
  if self._calc_info(blas):
blas_info:
  libraries blas not found in ['C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\lib', 'C:\\', 'C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\libs', 'C:\\Users\\nikol\\anaconda3\\Library\\lib']
  NOT AVAILABLE

C:\Users\nikol\OneDrive\Documents\OpenProjects\Python\numpy\numpy\distutils\system_info.py:2026: UserWarning: 
    Blas (http://www.netlib.org/blas/) libraries not found.
    Directories to search for the libraries can be specified in the
    numpy/distutils/site.cfg file (section [blas]) or by setting
    the BLAS environment variable.
  if self._calc_info(blas):
blas_src_info:
  NOT AVAILABLE

C:\Users\nikol\OneDrive\Documents\OpenProjects\Python\numpy\numpy\distutils\system_info.py:2026: UserWarning: 
    Blas (http://www.netlib.org/blas/) sources not found.
    Directories to search for the sources can be specified in the
    numpy/distutils/site.cfg file (section [blas_src]) or by setting
    the BLAS_SRC environment variable.
  if self._calc_info(blas):
  NOT AVAILABLE

non-existing path in 'numpy\\distutils': 'site.cfg'
lapack_opt_info:
lapack_mkl_info:
  libraries mkl_rt not found in ['C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\lib', 'C:\\', 'C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\libs', 'C:\\Users\\nikol\\anaconda3\\Library\\lib']
  NOT AVAILABLE

openblas_lapack_info:
  libraries openblas not found in ['C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\lib', 'C:\\', 'C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\libs', 'C:\\Users\\nikol\\anaconda3\\Library\\lib']
get_default_fcompiler: matching types: '['gnu', 'intelv', 'absoft', 'compaqv', 'intelev', 'gnu95', 'g95', 'intelvem', 'intelem', 'flang']'
customize GnuFCompiler
customize IntelVisualFCompiler
customize AbsoftFCompiler
customize CompaqVisualFCompiler
customize IntelItaniumVisualFCompiler
customize Gnu95FCompiler
customize G95FCompiler
customize IntelEM64VisualFCompiler
customize IntelEM64TFCompiler
customize PGroupFlangCompiler
  NOT AVAILABLE

openblas_clapack_info:
  libraries openblas,lapack not found in ['C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\lib', 'C:\\', 'C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\libs', 'C:\\Users\\nikol\\anaconda3\\Library\\lib']
get_default_fcompiler: matching types: '['gnu', 'intelv', 'absoft', 'compaqv', 'intelev', 'gnu95', 'g95', 'intelvem', 'intelem', 'flang']'
customize GnuFCompiler
customize IntelVisualFCompiler
customize AbsoftFCompiler
customize CompaqVisualFCompiler
customize IntelItaniumVisualFCompiler
customize Gnu95FCompiler
customize G95FCompiler
customize IntelEM64VisualFCompiler
customize IntelEM64TFCompiler
customize PGroupFlangCompiler
  NOT AVAILABLE

flame_info:
  libraries flame not found in ['C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\lib', 'C:\\', 'C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\libs', 'C:\\Users\\nikol\\anaconda3\\Library\\lib']
  NOT AVAILABLE

atlas_3_10_threads_info:
Setting PTATLAS=ATLAS
  libraries lapack_atlas not found in C:\Users\nikol\anaconda3\envs\numpy-dev\lib
  libraries tatlas,tatlas not found in C:\Users\nikol\anaconda3\envs\numpy-dev\lib
  libraries lapack_atlas not found in C:\
  libraries tatlas,tatlas not found in C:\
  libraries lapack_atlas not found in C:\Users\nikol\anaconda3\envs\numpy-dev\libs
  libraries tatlas,tatlas not found in C:\Users\nikol\anaconda3\envs\numpy-dev\libs
  libraries lapack_atlas not found in C:\Users\nikol\anaconda3\Library\lib
  libraries tatlas,tatlas not found in C:\Users\nikol\anaconda3\Library\lib
<class 'numpy.distutils.system_info.atlas_3_10_threads_info'>
  NOT AVAILABLE

atlas_3_10_info:
  libraries lapack_atlas not found in C:\Users\nikol\anaconda3\envs\numpy-dev\lib
  libraries satlas,satlas not found in C:\Users\nikol\anaconda3\envs\numpy-dev\lib
  libraries lapack_atlas not found in C:\
  libraries satlas,satlas not found in C:\
  libraries lapack_atlas not found in C:\Users\nikol\anaconda3\envs\numpy-dev\libs
  libraries satlas,satlas not found in C:\Users\nikol\anaconda3\envs\numpy-dev\libs
  libraries lapack_atlas not found in C:\Users\nikol\anaconda3\Library\lib
  libraries satlas,satlas not found in C:\Users\nikol\anaconda3\Library\lib
<class 'numpy.distutils.system_info.atlas_3_10_info'>
  NOT AVAILABLE

atlas_threads_info:
Setting PTATLAS=ATLAS
  libraries lapack_atlas not found in C:\Users\nikol\anaconda3\envs\numpy-dev\lib
  libraries ptf77blas,ptcblas,atlas not found in C:\Users\nikol\anaconda3\envs\numpy-dev\lib
  libraries lapack_atlas not found in C:\
  libraries ptf77blas,ptcblas,atlas not found in C:\
  libraries lapack_atlas not found in C:\Users\nikol\anaconda3\envs\numpy-dev\libs
  libraries ptf77blas,ptcblas,atlas not found in C:\Users\nikol\anaconda3\envs\numpy-dev\libs
  libraries lapack_atlas not found in C:\Users\nikol\anaconda3\Library\lib
  libraries ptf77blas,ptcblas,atlas not found in C:\Users\nikol\anaconda3\Library\lib
<class 'numpy.distutils.system_info.atlas_threads_info'>
  NOT AVAILABLE

atlas_info:
  libraries lapack_atlas not found in C:\Users\nikol\anaconda3\envs\numpy-dev\lib
  libraries f77blas,cblas,atlas not found in C:\Users\nikol\anaconda3\envs\numpy-dev\lib
  libraries lapack_atlas not found in C:\
  libraries f77blas,cblas,atlas not found in C:\
  libraries lapack_atlas not found in C:\Users\nikol\anaconda3\envs\numpy-dev\libs
  libraries f77blas,cblas,atlas not found in C:\Users\nikol\anaconda3\envs\numpy-dev\libs
  libraries lapack_atlas not found in C:\Users\nikol\anaconda3\Library\lib
  libraries f77blas,cblas,atlas not found in C:\Users\nikol\anaconda3\Library\lib
<class 'numpy.distutils.system_info.atlas_info'>
  NOT AVAILABLE

lapack_info:
  libraries lapack not found in ['C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\lib', 'C:\\', 'C:\\Users\\nikol\\anaconda3\\envs\\numpy-dev\\libs', 'C:\\Users\\nikol\\anaconda3\\Library\\lib']
  NOT AVAILABLE

C:\Users\nikol\OneDrive\Documents\OpenProjects\Python\numpy\numpy\distutils\system_info.py:1858: UserWarning: 
    Lapack (http://www.netlib.org/lapack/) libraries not found.
    Directories to search for the libraries can be specified in the
    numpy/distutils/site.cfg file (section [lapack]) or by setting
    the LAPACK environment variable.
  return getattr(self, '_calc_info_{}'.format(name))()
lapack_src_info:
  NOT AVAILABLE

C:\Users\nikol\OneDrive\Documents\OpenProjects\Python\numpy\numpy\distutils\system_info.py:1858: UserWarning: 
    Lapack (http://www.netlib.org/lapack/) sources not found.
    Directories to search for the sources can be specified in the
    numpy/distutils/site.cfg file (section [lapack_src]) or by setting
    the LAPACK_SRC environment variable.
  return getattr(self, '_calc_info_{}'.format(name))()
  NOT AVAILABLE

numpy_linalg_lapack_lite:
  FOUND:
    language = c
    define_macros = [('HAVE_BLAS_ILP64', None), ('BLAS_SYMBOL_SUFFIX', '64_')]

Warning: attempted relative import with no known parent package
C:\Users\nikol\anaconda3\envs\numpy-dev\lib\distutils\dist.py:274: UserWarning: Unknown distribution option: 'define_macros'
  warnings.warn(msg)
running build
running config_cc
unifing config_cc, config, build_clib, build_ext, build commands --compiler options
running config_fc
unifing config_fc, config, build_clib, build_ext, build commands --fcompiler options
running build_src
build_src
building py_modules sources
creating build\src.win-amd64-3.9
creating build\src.win-amd64-3.9\numpy
creating build\src.win-amd64-3.9\numpy\distutils
building library "npymath" sources
LINK : fatal error LNK1104: cannot open file 'MSVCRT.lib'
Traceback (most recent call last):
  File "C:\Users\nikol\OneDrive\Documents\OpenProjects\Python\numpy\setup.py", line 447, in <module>
    setup_package()
  File "C:\Users\nikol\OneDrive\Documents\OpenProjects\Python\numpy\setup.py", line 439, in setup_package
    setup(**metadata)
  File "C:\Users\nikol\OneDrive\Documents\OpenProjects\Python\numpy\numpy\distutils\core.py", line 169, in setup
    return old_setup(**new_attr)
  File "C:\Users\nikol\anaconda3\envs\numpy-dev\lib\site-packages\setuptools\__init__.py", line 163, in setup
    return distutils.core.setup(**attrs)
  File "C:\Users\nikol\anaconda3\envs\numpy-dev\lib\distutils\core.py", line 148, in setup
    dist.run_commands()
  File "C:\Users\nikol\anaconda3\envs\numpy-dev\lib\distutils\dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "C:\Users\nikol\anaconda3\envs\numpy-dev\lib\distutils\dist.py", line 985, in run_command
    cmd_obj.run()
  File "C:\Users\nikol\OneDrive\Documents\OpenProjects\Python\numpy\numpy\distutils\command\build.py", line 61, in run
    old_build.run(self)
  File "C:\Users\nikol\anaconda3\envs\numpy-dev\lib\distutils\command\build.py", line 135, in run
    self.run_command(cmd_name)
  File "C:\Users\nikol\anaconda3\envs\numpy-dev\lib\distutils\cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "C:\Users\nikol\anaconda3\envs\numpy-dev\lib\distutils\dist.py", line 985, in run_command
    cmd_obj.run()
  File "C:\Users\nikol\OneDrive\Documents\OpenProjects\Python\numpy\numpy\distutils\command\build_src.py", line 144, in run
    self.build_sources()
  File "C:\Users\nikol\OneDrive\Documents\OpenProjects\Python\numpy\numpy\distutils\command\build_src.py", line 155, in build_sources
    self.build_library_sources(*libname_info)
  File "C:\Users\nikol\OneDrive\Documents\OpenProjects\Python\numpy\numpy\distutils\command\build_src.py", line 288, in build_library_sources
    sources = self.generate_sources(sources, (lib_name, build_info))
  File "C:\Users\nikol\OneDrive\Documents\OpenProjects\Python\numpy\numpy\distutils\command\build_src.py", line 378, in generate_sources
    source = func(extension, build_dir)
  File "numpy\core\setup.py", line 666, in get_mathlib_info
    raise RuntimeError("Broken toolchain: cannot link a simple C program")
RuntimeError: Broken toolchain: cannot link a simple C program
```

</details>

@theoniko you can include it directly here, and if it's long then enclose it in:
```
<details>

...

</details>
```
Much easier than letting us download build logs.

--- a/numpy/core/src/multiarray/number.c
+++ b/numpy/core/src/multiarray/number.c
@@ -429,6 +429,7 @@ is_scalar_with_conversion(PyObject *o2, double* out_exponent)
             return NPY_NOSCALAR;
         }
         val = PyLong_AsSsize_t(value);
+        Py_DECREF(value);
         if (error_converting(val)) {
             PyErr_Clear();
             return NPY_NOSCALAR;

